File: code/day3.php
<?php
include 'file-handle.php';

function priority($char)
{
    return ctype_lower($char) ?
        (ord($char) - ord('a') + 1) : (ord($char) - ord('A') + 27);
}

function puzzle5($input_file)
{
    $data = trim(read_file($input_file));
    $rucksacks = preg_split('/\r?\n/', $data);
    $intersections = [];
    foreach ($rucksacks as $rucksack) {
        $middle = strlen($rucksack) / 2;
        $part1 = substr($rucksack, 0, $middle);
        $part2 = substr($rucksack, $middle, $middle);
        for ($i = 0; $i < strlen($part1); $i++)
            if (str_contains($part2, $part1[$i])) {
                array_push($intersections, $part1[$i]);
                break;
            }
    }
    $priorities = array_map('priority', $intersections);
    return array_sum($priorities);
}


function puzzle6($input_file)
{
    $data = trim(read_file($input_file));
    $rucksacks = preg_split('/\r?\n/', $data);
    $intersections = [];
    for ($i = 0; $i < sizeof($rucksacks); $i += 3)
        for ($j = 0; $j < strlen($rucksacks[$i]); ++$j) {
            $ch = $rucksacks[$i][$j];
            if (str_contains($rucksacks[$i + 1], $ch) && str_contains($rucksacks[$i + 2], $ch)) {
                array_push($intersections, $ch);
                break;
            }
        }
    $priorities = array_map('priority', $intersections);
    return array_sum($priorities);
}

echo 'Day3, part one: ' . puzzle5('./input/day3.txt') . PHP_EOL;
echo 'Day3, part two: ' . puzzle6('./input/day3.txt') . PHP_EOL;

File: code/day3.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <cctype>
#include <vector>
#include <regex>
#include <numeric>
using namespace std;

string read_file(const string &input_file)
{
    ifstream file(input_file);
    stringstream buffer;
    buffer << file.rdbuf();
    file.close();
    return buffer.str();
}

string rtrim(string &str)
{
    int last = str.find_last_not_of(" \t\r\n");
    str.erase(last + 1);
    return str;
}

vector<string> split(const string &str, const regex &delimiter)
{
    sregex_token_iterator itr(str.begin(), str.end(), delimiter, -1);
    const sregex_token_iterator end;
    vector<string> parts(itr, end);
    return parts;
}

int priority(const char ch)
{
    return islower(ch) ? (int(ch) - int('a') + 1) : (int(ch) - int('A') + 27);
}

int puzzle5(string input_file)
{
    string data = read_file(input_file);
    data = rtrim(data);
    vector<string> rucksacks = split(data, regex("\r?\n"));
    vector<char> intersections;
    int middle;
    string part1, part2;
    for (const string rucksack : rucksacks)
    {
        middle = rucksack.length() / 2;
        part1 = rucksack.substr(0, middle);
        part2 = rucksack.substr(middle, middle);
        for (char ch : part1)
            if (part2.find(ch) != string::npos)
            {
                intersections.push_back(ch);
                break;
            }
    }
    int sum = accumulate(intersections.begin(), intersections.end(), 0,
                         [](int previousResult, char ch)
                         { return previousResult + priority(ch); });
    return sum;
}

int puzzle6(string input_file)
{
    string data = read_file(input_file);
    data = rtrim(data);
    vector<string> rucksacks = split(data, regex("\r?\n"));
    vector<char> intersections;
    string rucksack1, rucksack2, rucksack3;
    for (int i = 0; i < rucksacks.size(); i += 3)
    {
        for (char ch : rucksacks[i])
            if (rucksacks[i + 1].find(ch) != string::npos &&
                rucksacks[i + 2].find(ch) != string::npos)
            {
                intersections.push_back(ch);
                break;
            }
    }
    int sum = accumulate(intersections.begin(), intersections.end(), 0,
                         [](int previousResult, char ch)
                         { return previousResult + priority(ch); });
    return sum;
}

int main(void)
{
    cout << "Day #3, part one: " << puzzle5("input/day3.txt") << endl;
    cout << "Day #3, part two: " << puzzle6("input/day3.txt") << endl;
}